lnwallet/channel.go

```diff
@@ -894,8 +931,12 @@ func (lc *LightningChannel) diskHtlcToPayDesc(feeRate chainfee.SatPerKWeight,
 
 	// Ensure that we'll restore any custom records which were stored as
 	// extra data on disk.
-	if len(htlc.ExtraData) != 0 {
-		pd.CustomRecords = fn.Some[tlv.Blob](htlc.ExtraData)
+	if len(htlc.CustomRecords) != 0 {
```

Comment:
While we're setting the `CustomRecords` values [above](https://github.com/lightningnetwork/lnd/pull/8660/commits/9f15cb3019ce08abfd3b26eb6e54b416c6fb1864#diff-cdf9f47cef64ceac10ee55ed7bf0569ac54f5d5f1af599ef51ccaf5db813d03eR824), we're not actually persisting them using the workaround added in #7710, which means that in certain restart cases we won't restore these records (diff to test db restore
[here](https://github.com/lightningnetwork/lnd/files/15377411/persist_htlc.patch)).

So as-is we'll always have `htlc.CustomRecords`=0 because we're not persisting them with the `channeldb.HTLC` that needs to be turned into a `PaymentDescriptor` here. 

The restart flow always takes me forever to reload into my brain, tried to write it all down [here](https://docs.google.com/document/d/1ZbMxZ9ao1ekHXUjJusTSTvZsYciZ1CzvZbK56CiDBEI/edit) if that's at all helpful.

Reply:
Right, persistence is missing atm.

So I'm not sure if the best way forward is persisting this as part of `channeldb.HTLC` or `channeldb.PaymentCreationInfo` as suggested [here](https://github.com/lightningnetwork/lnd/pull/8660#discussion_r1577161403)

Reply:
> So I'm not sure if the best way forward is persisting this as part of channeldb.HTLC or channeldb.PaymentCreationInfo as suggested https://github.com/lightningnetwork/lnd/pull/8660#discussion_r1577161403

We need persistence in `channeldb.HTLC` so that the node that receives the incoming HTLC with custom TLVs is robust against restarts:
1. HTLC is irrevocably committed on the channel
2. Receiving node restarts before forwarding through the switch
3. HTLC is restored from disk (no custom records)
4. Interceptor receives HTLC without custom records